src/hotspot/share/cds/cdsConstants.cpp

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2014, 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include <cstddef>
```

Comment:
It is not stated in the style guide, but I think most files include the system headers after the HotSpot headers, separated by a blank line.

Reply:
Is this even needed now? I don't see it being used.

Reply:
That change wasn't what I requested. They layout is:
```
#include "precompiled.hpp"
#include "cds.hpp"
...
#include "cds/filemap.hpp"

#include <cstddef>
```

However, I think the more appropriate change is to include globalDefinitions.hpp instead of explicitly including cstddef.

Reply:
The pre-submit tests failed to build hotspot for linux on arm/ppc/s390 etc after <cstddef> removed. <cstddef> needed for defining size_t. So will add it back at last section of 'include'.

Reply:
The problem is here in cdsConstants.hpp

```
#include "memory/allStatic.hpp"
typedef struct {
  const char* _name;
  size_t _value;
} CDSConst;
```

globalDefinitions.hpp should be included here for the declaration of `size_t`.

Reply:
Yes, but it still failed for non-pch since cds.h is before cdsConstants.hpp. size_t in cds.h is not defined yet. Should I add it to cds.h?

Reply:
This is a bit complicated. The file is src/hotspot/share/include/cds.h, and the HotSpot convention for including files in this directory usually drops the "include/" part. E.g.,

```
#include "jvm.h"
```

But this will put `#include "cds.h"` at the very beginning of all includes, so it doesn't take up `size_t` which is indirectly declared later via globalDefinitions.hpp.

However, header files in src/hotspot/share/ may be included by C/C++ files outside of HotSpot. As a result, these headers cannot include other HotSpot headers. I.e., you cannot put `#include "utilities/globalDefinitions.hpp"` in cds.h, because cds.h is included by src/jdk.hotspot.agent/share/native/libsaproc/ps_core_common.c.

I think one compromise is to add `#include <stddefs.h>` in cds.h. Since cds.h can be included by C source code, we cannot use `#include <cstddef>` which is a C++ thing.  `<stddefs.h>` is part of ANSI C standard and we already include it inside share/utilities/debug.hpp, so we can safely assume that it exists for all compilers that can build HotSpot.

Reply:
I agree with Ioi - use stddefs.h in cds.h